### PR TITLE
Testing feature

### PR DESCRIPTION
**Pull request by Armynerh**
@fatmahussein  and @Armynerh implemented the following requirements:
We set up the React Testing Library.
We followed the instructions in the official [documentation](https://testing-library.com/docs/react-testing-library/intro).
We wrote unit tests for the files operate.js and calculate.js using Jest.
We created unit tests for all React components:
We used Jest snapshots to test the components.
We used the React Testing Library to simulate user interaction.
We ran full tests suite using Jest. All tests pass.

test

### CODE REVIEW

File: src/logic/calculate.test.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
+import calculate from './calculate';
```
Comment (ShadyShawkat) [inline src/logic/calculate.test.js:1]: - Good job implementing tests for the "calculate.js" and "operate.js" files. 👍🏻 
- [ ] Kindly implement more tests that use Jest snapshots for testing all the components and tests that use the React Testing Library to simulate user interaction. As per [these](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/project_tests.md#:~:text=Use%20Jest%20snapshots,simulate%20user%20interaction.) requirements.

**Review by ShadyShawkat — CHANGES_REQUESTED** [review]
## Hi @Armynerh, 

Good job so far! 👍🏻 
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight
- Correct use of GitFlow ✔️ 
- Descriptive commit messages ✔️ 
- Professional looking pull request ✔️ 

Image

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/input.jsx
Comment (shahadat3669) [inline src/components/input.jsx:1]: - [ ] Well! You've evidently made strides in incorporating Jest snapshots for many components. However, in alignment with the [project requirement](https://github.com/microverseinc/curriculum-react-redux/blob/main/math-magicians/project_tests.md#:~:text=Create%20unit%20tests,simulate%20user%20interaction.), your testing scope should encompass all components. This comprehensive approach ensures robustness and reliability.

Kindly adhere to the project's specific requirements and extend your testing suite to encompass every file. This meticulous effort will fulfill the criteria and fortify your application against potential issues. Keep up the diligent work! 🧪👨‍🔬📊

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @Armynerh, @fatmahussein, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ No linter Errors
✅ Usage of Jest snapshots for testing
✅ Usage of React Testing Library to simulate user interaction



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table  width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/components/input.jsx
Comment (fatmahussein) [reply src/components/input.jsx:1]: @shahadat3669 could you please elaborate

**Review by medaminedev66 — COMMENTED** [review]
Hello @fatmahussein👋,

_This is an additional review as a result of the second opinion request._

According to the project requirements, you should provide snapshot tests for all the components you've created in this project. For example, the Button component should have a corresponding snapshot test.

Keep rocking 🚀!

**Comment by Armynerh** [comment]
Hi @medaminedev66 
You gave a second opinion on our project for the math magician test. I would love it if you could please help us take a look at it and tell us if it’s okay to ask for a final review. the changes required have been implemented. Thank you

**Comment by shahadat3669** [comment]
> Hi @medaminedev66 You gave a second opinion on our project for the math magician test. I would love it if you could please help us take a look at it and tell us if it’s okay to ask for a final review. the changes required have been implemented. Thank you

Hello @fatmahussein,

I apologize for overlooking your previous mention. I'm sorry to inform you that there is files that still requires testing: for example [fetchQuote.jsx](https://github.com/fatmahussein/Math-Magicians/blob/testing-feature/src/components/fetchQuote.jsx). Could you please include test cases for this file as well? Your attention to this matter is greatly appreciated.

File: src/components/input.jsx
Comment (peterrobert) [reply src/components/input.jsx:1]: This change has been done well

**Review by peterrobert — APPROVED** [review]
Hi @Armynerh , 


### Highlights!

[ ] Good feature branch name ✔️
[ ] Descriptive pull request ✔️
[ ] Great work adding tests and snapshots for each component in your application✔️


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread or tag me @peterrobert if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**